### PR TITLE
Move uhttp module to dig

### PR DESCRIPTION
**Pull request by alsamylkin**
Refactor uhttp module to implement to fx.Module interface:
* Module constructor accepts a component that should return an http handler.
* removed service.Host from everywhere in modules/uhttp.
* updated docs to show how new handler constructor can accept a config.
* added more tests.

### CODE REVIEW

**Comment by alsamylkin** [comment]
Do I need to update the changelog?:)

File: modules/uhttp/http.go [hunk 1]
```diff
@@ -29,122 +29,111 @@ import (
...
-func newModule(host service.Host, handler http.Handler) (*Module, error) {
-	// setup config defaults
-	cfg := Config{
-		Port:    defaultPort,
-		Timeout: defaultTimeout,
-	}
-	log := ulog.Logger(context.Background()).With(zap.String("module", host.ModuleName()))
-	if err := host.Config().Get("modules").Get(host.ModuleName()).Populate(&cfg); err != nil {
-		log.Error("Error loading http module configuration", zap.Error(err))
+func New(handlerCtor fx.Component) *Module {
```
Comment (alsamylkin) [inline modules/uhttp/http.go]: I would really prefer to have it as an abstract function, that returns modules, we can pass configs/loggers/etc directly to modules then.

File: modules/uhttp/http.go [hunk 2]
```diff
@@ -29,122 +29,111 @@ import (
...
-	}, nil
+// Name returns uhttp.
+func (m *Module) Name() string {
+	return "uhttp"
 }
 
-// Start begins serving requests over HTTP
-func (m *Module) Start() error {
-	return nil
+type starter struct{}
```
Comment (anuptalwalkar) [inline modules/uhttp/http.go]: Why do you need empty starter struct? This is going to be a module construct, you can just return error.

File: modules/uhttp/http.go [hunk 3]
```diff
@@ -29,122 +29,111 @@ import (
...
+
+// Constructor returns module components: handler constructor and harness for it.
+func (m *Module) Constructor() []fx.Component {
+	return []fx.Component{
+		m.handlerCtor,
+		func(provider config.Provider, l *zap.Logger, scope tally.Scope, handler http.Handler) (*starter, error) {
+			// setup config defaults
+			cfg := Config{}
+
+			m.l = ulog.Logger(context.Background()).With(zap.String("module", m.Name()))
```
Comment (anuptalwalkar) [inline modules/uhttp/http.go]: `m.l = l.With(zap.String("module", m.Name())` ?

File: modules/uhttp/http_test.go [hunk 4]
```diff
@@ -79,48 +91,81 @@ func TestHTTPModule_Tracer(t *testing.T) {
...
-	)
-	require.NoError(t, err)
-	mod, err := newModule(host, handler)
-	if expectError {
-		require.Error(t, err, "Expected error instantiating module")
-		fn(nil)
-		return
+	di := dig.New()
+	p := config.NewYAMLProviderFromBytes(_httpConfig)
+	di.MustRegister(&p)
```
Comment (anuptalwalkar) [inline modules/uhttp/http_test.go]: you can use `di.ProvideAll`.. which returns an err

File: modules/uhttp/http_test.go [hunk 5]
```diff
@@ -79,48 +91,81 @@ func TestHTTPModule_Tracer(t *testing.T) {
...
+	mod := New(&handler)
+	ctors := mod.Constructor()
+	for i := range ctors {
+		di.MustRegister(ctors[i])
 	}
-	require.NoError(t, err, "Unable to instantiate module")
-	assert.NoError(t, mod.Start(), "Got error from starting")
+
+	var s *starter
+	di.MustResolve(&s)
```
Comment (anuptalwalkar) [inline modules/uhttp/http_test.go]: `Resolve` and check for error

File: modules/uhttp/http_test.go [hunk 6]
```diff
@@ -79,48 +91,81 @@ func TestHTTPModule_Tracer(t *testing.T) {
...
+	di := dig.New()
+	p := config.NewYAMLProviderFromBytes(_httpConfig)
+	di.MustRegister(&p)
+	di.MustRegister(&tally.NoopScope)
+	di.MustRegister(zap.NewNop())
+
+	mod := New(&handler)
+	ctors := mod.Constructor()
+	for i := range ctors {
+		di.MustRegister(ctors[i])
```
Comment (anuptalwalkar) [inline modules/uhttp/http_test.go]: `err := di.Invoke(ctors[i])` and you don't need to resolve starter below. :)

File: modules/uhttp/http.go (hunk 2 above)
Comment (alsamylkin) [inline modules/uhttp/http.go]: Oh yes, Service, just calls Invoke()

**Pull request merged**